packages/autocomplete-js/src/types/AutocompleteSource.ts

```diff
@@ -4,11 +4,16 @@ import {
   BaseItem,
 } from '@algolia/autocomplete-core';
 
+import { AutocompleteComponents } from './AutocompleteComponents';
 import { Pragma, PragmaFrag, VNode } from './AutocompleteRenderer';
 import { AutocompleteState } from './AutocompleteState';
 
 type Template<TParams> = (
-  params: TParams & { createElement: Pragma; Fragment: PragmaFrag }
+  params: TParams & {
+    createElement: Pragma;
+    Fragment: PragmaFrag;
+    components: AutocompleteComponents;
```

Comment:
could there not be custom components here?

Reply:
It could—the type supports user-provided component:

https://github.com/algolia/autocomplete/blob/03134554e97fb2ddc8e67e991ebba0dbd86ef513/packages/autocomplete-js/src/types/AutocompleteComponents.ts#L14-L19

Reply:
ahh because users should override this type?

Reply:
They don't have to, but they could extend the `AutocompleteComponents` interface if they want proper typing on their custom components.